### PR TITLE
Addition of upstream proxy to resolve slice redirect errors (302 Playstation)

### LINKED ISSUE #60: Playstation 302 Header Errors
### Describe the issue you are having
```
Receiving "unexpected status code 302 in slice response while reading response header from upstream" errors on playstation caching.
```

### How are you running the container(s)?
```
docker run --restart unless-stopped --name lancache-dns --detach -p 53:53/udp -e USE_GENERIC_CACHE=true -e LANCACHE_IP=192.168.0.25 -e UPSTREAM_DNS="9.9.9.9;149.112.112.112" lancachenet/lancache-dns:latest
docker run --restart unless-stopped --name lancache --detach -v /home/user/cache/data:/data/cache -v /mnt/logs:/data/logs -p 80:80 -e UPSTREAM_DNS="9.9.9.9 149.112.112.112"  lancachenet/monolithic:latest
docker run --restart unless-stopped --name lancache-sniproxy --detach -p 443:443 -e UPSTREAM_DNS="9.9.9.9 149.112.112.112" lancachenet/sniproxy:latest
```

### DNS Configuration
```
Router is directing all systems to the DNS.
```
### Output of container(s)
```
2019/10/02 08:37:59 [error] 1817#1817: *2484 unexpected status code 302 in slice response while reading response header from upstream, client: 192.168.0.226, server: , request: "GET /gs2/ppkgo/prod/CUSA07022_00/178/f_25375fa9ae011b3580d173f79416e588b81c8f20123de80188ea6a75c0612b1a/f/UP1477-CUSA07022_00-FORTNITETESTING1-A0238-V0100_0.pkg?downloadId=00000e38&du=0000000000000e3800e253e85c6e4184&product=0184&threadId=0&serverIpAddr=192.168.0.25&r=00000001&downloadType=om HTTP/1.1", subrequest: "/gs2/ppkgo/prod/CUSA07022_00/178/f_25375fa9ae011b3580d173f79416e588b81c8f20123de80188ea6a75c0612b1a/f/UP1477-CUSA07022_00-FORTNITETESTING1-A0238-V0100_0.pkg", upstream: "http://69.28.165.0:80/gs2/ppkgo/prod/CUSA07022_00/178/f_25375fa9ae011b3580d173f79416e588b81c8f20123de80188ea6a75c0612b1a/f/UP1477-CUSA07022_00-FORTNITETESTING1-A0238-V0100_0.pkg?downloadId=00000e38&du=0000000000000e3800e253e85c6e4184&product=0184&threadId=0&serverIpAddr=192.168.0.25&r=00000001&downloadType=om", host: "gs2.ww.prod.dl.playstation.net"

```

### PR DESCRIPTION
**Pull request by VibroAxe**
Added new upstream proxy in lancache standard structure.

This resolves #60 by sending all upstream requests via a small 2nd proxy to catch any 302 responses coming from the load balancer "mid slice". Any 302 requests are first followed by the interim proxy and then re-served (slices intact) to the primary lancache block to be cached.

For reasons best known to past me, we seem to be getting some duplicated commits in this PR which are already in main. But all seem to be operational (assuming the CI passes).

### CODE REVIEW

**Comment by mintopia** [comment]
lgtm

File: overlay/etc/nginx/nginx.conf [hunk 1]
```diff
@@ -25,6 +25,7 @@ http {
 	error_log /var/log/nginx/error.log;
 
 
+
```
Comment (MathewBurnett) [inline overlay/etc/nginx/nginx.conf:28]: new line for effect

File: overlay/etc/nginx/sites-available/20_upstream.conf [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Upstream server to proxy and handle inconsistent 302 redirects
+# All cache traffic is passed through this proxy to allow rewriting of redirects without caching
+
+# This is particularly important for sony / ps5 as upstreams redirect between them which confuses slice map on caching
+
+server {
+
+	# Internal bind on 3128, this should not be externally mapped
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/20_upstream.conf:8]: odd indent

File: overlay/etc/nginx/sites-available/20_upstream.conf [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  listen localhost:3128 reuseport;
+
+	# No access_log tracking as all requests to this instance are already logged through monolithic
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/20_upstream.conf:11]: odd indent

File: overlay/etc/nginx/sites-available/cache.conf.d/root/90_upstream.conf [hunk 4]
```diff
@@ -1,6 +1,10 @@
     # Upstream Configuration
     proxy_next_upstream error timeout http_404;
-    proxy_pass http://$host$request_uri;
+#    proxy_pass http://$host$request_uri;
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/cache.conf.d/root/90_upstream.conf:3]: if this is going, loose the line

File: overlay/etc/nginx/sites-available/cache.conf.d/root/90_upstream.conf [hunk 5]
```diff
@@ -1,6 +1,10 @@
...
+
+		# Proxy into the redirect handler
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/cache.conf.d/root/90_upstream.conf:5]: odd indent

File: overlay/etc/nginx/sites-available/upstream.conf.d/30_primary_proxy.conf [hunk 6]
```diff
@@ -0,0 +1,9 @@
+	# Proxy all requests to upstream
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/upstream.conf.d/30_primary_proxy.conf:1]: odd indent

File: overlay/etc/nginx/sites-available/upstream.conf.d/30_primary_proxy.conf [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+  location / {
+    # Simple proxy the request
+		proxy_pass http://$host$request_uri;
+
+		# Catch the errors to process the redirects
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/upstream.conf.d/30_primary_proxy.conf:6]: odd indent

File: overlay/etc/nginx/sites-available/upstream.conf.d/20_tracking.conf [hunk 8]
```diff
@@ -0,0 +1,2 @@
+	# Header to track if resolved from upstream or 302 redirect
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/upstream.conf.d/20_tracking.conf:1]: odd indent

File: overlay/etc/nginx/sites-available/upstream.conf.d/40_redirect_proxy.conf [hunk 9]
```diff
@@ -0,0 +1,11 @@
+	# Special location block to handle 302 redirects
+  location @upstream_redirect {
+		# Upstream_http_location contains the Location: redirection from the upstream server
```
Comment (MathewBurnett) [inline overlay/etc/nginx/sites-available/upstream.conf.d/40_redirect_proxy.conf:3]: odd indent

**Review by MathewBurnett — CHANGES_REQUESTED** [review]
mostly ok, some odd indent errors. Also contains
* Moving apt update in docker file
* Cache mem size update
* ipv6 resolver off
* clacks header

File: Dockerfile [hunk 10]
```diff
@@ -17,6 +17,9 @@ ENV GENERICCACHE_VERSION=2 \
     CACHE_DOMAINS_BRANCH=master \
     NGINX_WORKER_PROCESSES=auto
 
+RUN	apt-get update							;\
```
[the remark this answers is not in the document]
Comment (VibroAxe) [reply Dockerfile:20]: Which other bit?

**Comment by alexdrean** [comment]
I would love to see this in master so I fixed the indent issues... 
https://github.com/lancachenet/monolithic/compare/playstation...alexdrean:monolithic:playstation

To apply to your local `playstation` branch:
```bash
git remote add alexdrean https://github.com/alexdrean/monolithic.git
git remote update
git checkout origin/playstation
git merge alexdrean/playstation
`


Also, I think Mathew was maybe trying to say that
`RUN apt-get update; install git jq`
should/could go right before
`RUN rm /etc/nginx/sites-enabled/* /etc/nginx/stream-enabled/* ;\`.
That's just a guess, I've been wrong before.

Thank you so much for doing all this debugging work! Fixed my customers' recent issues with PS4/5.

File: Dockerfile [hunk 11]
```diff
@@ -31,12 +34,11 @@ RUN rm /etc/nginx/sites-enabled/* /etc/nginx/stream-enabled/* ;\
...
 	mkdir -p /etc/nginx/sites-enabled	;\
 	ln -s /etc/nginx/sites-available/10_cache.conf /etc/nginx/sites-enabled/10_generic.conf; \
+	ln -s /etc/nginx/sites-available/20_upstream.conf /etc/nginx/sites-enabled/20_upstream.conf; \
 	ln -s /etc/nginx/stream-available/10_sni.conf /etc/nginx/stream-enabled/10_sni.conf
 
 RUN mkdir -m 755 -p /data/cachedomains		;\
-	mkdir -m 755 -p /tmp/nginx				;\
-	apt-get update							;\
-	apt-get install -y jq git				;
+	mkdir -m 755 -p /tmp/nginx
```
Comment (MathewBurnett) [inline Dockerfile:41]: Can this go in the layer with the other mkdir?

File: Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (VibroAxe) [reply Dockerfile]: Moved above ENV to improve flow

**Review by MathewBurnett — APPROVED** [review]

**Pull request merged**